tls/s2n_kem.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_kem.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_kem_generate_key_pair(const struct s2n_kem *kem, struct s2n_kem_params *kem_params)
+{
+    notnull_check(kem->generate_keypair);
+    GUARD(s2n_alloc(&kem_params->public_key, kem->publicKeySize));
+    GUARD(s2n_alloc(&kem_params->private_key, kem->privateKeySize));
+    GUARD(kem->generate_keypair(kem_params->public_key.data, kem_params->private_key.data));
+    return 0;
+}
+
+int s2n_kem_generate_shared_secret(const struct s2n_kem *kem, struct s2n_kem_params *kem_params,
+                                   struct s2n_blob *shared_secret, struct s2n_blob *ciphertext)
+{
+    notnull_check(kem->encrypt);
+    GUARD(s2n_alloc(shared_secret, kem->sharedSecretKeySize));
+    GUARD(s2n_alloc(ciphertext, kem->ciphertextSize));
+    GUARD(kem->encrypt(ciphertext->data, shared_secret->data, kem_params->public_key.data));
+    return 0;
+}
+
+int s2n_kem_decrypt_shared_secret(const struct s2n_kem *kem, struct s2n_kem_params *kem_params,
+                                  struct s2n_blob *shared_secret, struct s2n_blob *ciphertext)
+{
+    notnull_check(kem->decrypt);
+    notnull_check(kem_params->private_key.data);
+    eq_check(kem->ciphertextSize, ciphertext->size);
+
+    GUARD(s2n_alloc(shared_secret, kem->sharedSecretKeySize));
+    GUARD(kem->decrypt(shared_secret->data, ciphertext->data, kem_params->private_key.data));
```

Comment:
According to the spec:
"In the case of a decapsulation failure, an implementation MUST abort the handshake with a decrypt_error fatal alert"

Where do you send this error?

Reply:
If the KEM returns a failure the chain of GUARDs will fail back up to https://github.com/awslabs/s2n/blob/master/tls/s2n_handshake_io.c#L599 which will cause `s2n_negotiate` to fail. s2n does not send any message back to the client by design.

Reply:
I am working with mattcampagna on possibly updating the specification.